### PR TITLE
TST: Consolidate tests that raise in groupby

### PR DESCRIPTION
**Pull request by albuzenet**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Part of #50133. Continue the work of @rhshadrach started in #50404 for the groupby consolidation.

### CODE REVIEW

File: pandas/tests/groupby/aggregate/test_cython.py [hunk 1]
```diff
@@ -95,13 +95,8 @@ def test_cython_agg_nothing_to_agg():
     with pytest.raises(TypeError, match="Cannot use numeric_only=True"):
         frame.groupby("a")["b"].mean(numeric_only=True)
 
-    with pytest.raises(TypeError, match="Could not convert (foo|bar)*"):
-        frame.groupby("a")["b"].mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_cython.py:99]: This is not covered in the added tests

File: pandas/tests/groupby/aggregate/test_cython.py [hunk 2]
```diff
@@ -95,13 +95,8 @@ def test_cython_agg_nothing_to_agg():
...
-
     frame = DataFrame({"a": np.random.randint(0, 5, 50), "b": ["foo", "bar"] * 25})
 
-    with pytest.raises(TypeError, match="Could not convert"):
-        frame[["b"]].groupby(frame["a"]).mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/aggregate/test_cython.py:104]: This is not covered in the added tests

File: pandas/tests/groupby/test_categorical.py [hunk 3]
```diff
@@ -103,9 +103,6 @@ def test_basic():  # TODO: split this test
     gb = df.groupby("A", observed=False)
     exp_idx = CategoricalIndex(["a", "b", "z"], name="A", ordered=True)
     expected = DataFrame({"values": Series([3, 7, 0], index=exp_idx)})
-    msg = "category type does not support sum operations"
-    with pytest.raises(TypeError, match=msg):
-        gb.sum()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_categorical.py:108]: This is not covered in the added tests, can add `observed` to the test arguments

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -465,8 +465,6 @@ def test_multi_func(df):
     col2 = df["B"]
 
     grouped = df.groupby([col1.get, col2.get])
-    with pytest.raises(TypeError, match="Could not convert"):
-        grouped.mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:469]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 5]
```diff
@@ -663,17 +661,11 @@ def test_groupby_as_index_agg(df):
 
     # single-key
 
-    with pytest.raises(TypeError, match="Could not convert"):
-        grouped.agg(np.mean)
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:667]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 6]
```diff
@@ -958,11 +946,7 @@ def test_empty_groups_corner(mframe):
     )
 
     grouped = df.groupby(["k1", "k2"])
-    with pytest.raises(TypeError, match="Could not convert"):
-        grouped.agg(np.mean)
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:962]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 7]
```diff
@@ -958,11 +946,7 @@ def test_empty_groups_corner(mframe):
...
     result = grouped[["v1", "v2"]].agg(np.mean)
-    with pytest.raises(TypeError, match="Could not convert"):
-        grouped.mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:965]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 8]
```diff
@@ -1146,8 +1130,6 @@ def test_groupby_with_hier_columns():
     # add a nuisance column
     sorted_columns, _ = columns.sortlevel(0)
     df["A", "foo"] = "bar"
-    with pytest.raises(TypeError, match="Could not convert"):
-        df.groupby(level=0).mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:1150]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 9]
```diff
@@ -1181,11 +1163,7 @@ def test_groupby_wrong_multi_labels():
 
 
 def test_groupby_series_with_name(df):
-    with pytest.raises(TypeError, match="Could not convert"):
-        df.groupby(df["A"]).mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:1185]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 10]
```diff
@@ -1181,11 +1163,7 @@ def test_groupby_wrong_multi_labels():
...
     result = df.groupby(df["A"]).mean(numeric_only=True)
-    with pytest.raises(TypeError, match="Could not convert"):
-        df.groupby(df["A"], as_index=False).mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:1188]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 11]
```diff
@@ -1335,11 +1313,7 @@ def test_groupby_unit64_float_conversion():
 
 
 def test_groupby_list_infer_array_like(df):
-    with pytest.raises(TypeError, match="Could not convert"):
-        df.groupby(list(df["A"])).mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:1339]: This is not covered in the added tests

File: pandas/tests/groupby/test_groupby.py [hunk 12]
```diff
@@ -1335,11 +1313,7 @@ def test_groupby_unit64_float_conversion():
...
     result = df.groupby(list(df["A"])).mean(numeric_only=True)
-    with pytest.raises(TypeError, match="Could not convert"):
-        df.groupby(df["A"]).mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:1342]: This is not covered in the added tests

File: pandas/tests/groupby/test_index_as_string.py [hunk 13]
```diff
@@ -48,12 +48,7 @@ def series():
 )
 def test_grouper_index_level_as_string(frame, key_strs, groupers):
     if "B" not in key_strs or "outer" in frame.columns:
-        with pytest.raises(TypeError, match="Could not convert"):
-            frame.groupby(key_strs).mean()
         result = frame.groupby(key_strs).mean(numeric_only=True)
-
-        with pytest.raises(TypeError, match="Could not convert"):
-            frame.groupby(groupers).mean()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_index_as_string.py:56]: This is not covered in the added tests

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! This is a good step toward consolidating these. However there are many removed tests that are not part of the added tests (and conversely, there are many added tests that were not originally there! That bit is great). I've identified which I believe are missing below, please let me know if you think any are mistaken or would like explanation as to why.

I think if we can get coverage with the following cases, it's be good.

 - Multiple groupings (`.groupby(['a', 'b'])`)
 - observed=False for categorical
 - Series selection after groupby (`.groupby(...)['column']`)
 - Groupings that aren't in-axis (`.groupby([0, 0, 1])`)
 - Groupings that are groupers (`.groupby(pd.Grouper(...))`)
 - NumPy functions (`.groupby(...).agg(np.mean)`)

**Comment by albuzenet** [comment]
Thanks for the review @rhshadrach. 

I updated the tests, I believe now it should cover all the cases you mentioned (and more). Wondering if 28k tests for a single file isn't to much though ?

Anyway, I think we are still missing some cases like grouping on the index or grouping on a dtype other than int or category. Not sure if you think it is necessary here?

File: pandas/tests/groupby/test_raises.py [hunk 14]
```diff
@@ -4,23 +4,58 @@
...
+        Series([1, 1, 1, 1, 1, 2, 2, 2, 2]),
+        [Series([1, 1, 1, 1, 1, 2, 2, 2, 2]), Series([3, 3, 4, 4, 4, 4, 4, 3, 3])],
+    ]
+)
+def by(request):
+    return request.param
+
+
+@pytest.fixture(params=[True, False])
+def groupby_serie(request):
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_raises.py:38]: typo: groupby_series

File: pandas/tests/groupby/test_raises.py [hunk 15]
```diff
@@ -4,23 +4,58 @@
...
+            "d": list("xyzwtyuio"),
         }
     )
     args = get_groupby_method_args(groupby_func, df)
-    gb = df.groupby("a", as_index=as_index, sort=sort)
+    gb = df.groupby(by=by, as_index=as_index, sort=sort)
+
+    if groupby_serie:
+        if groupby_func == "corrwith":
+            pytest.skip()
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_raises.py:57]: Why skip here?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Wow - great work here! For any added tests, can you add `# GH#50749` to the top of any new tests.

What's the runtime of this file on your machine?

File: pandas/tests/groupby/test_raises.py (hunk 15 above)
Comment (albuzenet) [reply pandas/tests/groupby/test_raises.py:57]: It seems that the corrwith method is not implemented for SeriesGroupBy. We could test for an AttributeError, but it makes more sense to me to skip it.

**Comment by albuzenet** [comment]
On my machine, I get :
==== 28501 passed, 420 skipped in 113.16s (0:01:53) ====

Also, do you know why the CI pipeline is failing ? I see some failed tests in pandas/tests/indexes, but I don't see the connection with this PR ?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
This looks good. I am concerned though with the runtime for this. Can you remove `sort` and `as_index` from the tests - I think it's safe to test without these and would reduce the runtime by ~3/4ths.

> Also, do you know why the CI pipeline is failing ? I see some failed tests in pandas/tests, but I don't see the connection with this PR ?

This appears unrelated to me too. This is okay to ignore.

**Comment by albuzenet** [comment]
Done. The new runtime on my machine is : 
===== 10561 passed in 44.23s =====

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @albuzenet - great work!

**Comment by albuzenet** [comment]
Thanks for the review @rhshadrach !

**Pull request merged**